### ISSUE TITLE
LoadError on models in sub directories

### ISSUE THREAD

**Issue by gmoore** [wavii/rails-dev-tweaks#8]
We have a model in a subdirectory

```
/app/models/fa/base.rb
```

Defined like so

```
class FA::Base < ActiveResource::Base
  #code
end
```

We just upgraded from 3.0.9 -> 3.1.2.  The app works in 3.1.2.  I added this to our Gemfile

```
gem 'rails-dev-tweaks', '~> 0.6.1'
```

And then our app crashed on the first page load with

```
LoadError
Expected /Users/gary/dev/artful.ly/app/models/fa/base.rb to define Base
```

When we changed the class declaration to

```
class Base < ActiveResource::Base
```

Rails crashed with

```
ruby-1.9.2-p290/gems/activesupport-3.1.2/lib/active_support/dependencies.rb:490:in `load_missing_constant': Expected /Users/gary/dev/artful.ly/app/models/fa/base.rb to define FA::Base (LoadError)
```

Which was expected

Our autoload paths are:

```
config.autoload_paths += %W(#{config.root}/lib)
config.autoload_paths += Dir["#{config.root}/lib/**/"]
config.autoload_paths += Dir["#{Rails.root.to_s}/app/models/**/"]
```

I've been reading about autoload issues between 3.0 and 3.1, but again this app runs without rails-dev-tweaks installed.  Something is amiss somewhere.  I haven't dove into the rails-dev-tweaks code yet but I'm about to get going on it.

I tried this with 0.6.0 and 0.5.2

Using Ruby 1.9.2

Let me know what other info will help diagnose.  Thanks!

**Comment by nevir** [comment]
Hmm, [`rails-dev-tweaks` is relying on the autoloader](https://github.com/wavii/rails-dev-tweaks/blob/master/lib/rails-dev-tweaks.rb) for loading, but not reloading; I'm wondering if the paths are getting confused

A couple comments on your `config.autoload_paths` first though:
- You've got potentially conflicting paths going on with your `lib/` dir - `lib/foo/bar.rb` could match either `Foo::Bar` (your first config line) or `Bar` (your expanded 2nd line).
- By expanding your load paths, you have a very high likelihood of conflicting names.  `lib/base.rb` would conflict with `app/models/foo/base.rb` - you'd get the lib one first, I think.
- It's generally much safer to just specify the top level directories as autoloadable, and then rely on namespacing to take care of the rest

What does `ActiveSupport::Dependencies.autoload_paths` look like from a `rails console`?

**Comment by nevir** [comment]
In other words, try just this for your config:

```
config.autoload_paths << Rails.root.join('lib')
```

`app/models` is on the autoload paths by default; and by expanding it as you were, the autoloader was expecting `app/models/fa` as a top level namespace (thus why it's thinking `fa/base.rb` should define just `Base` instead of `FA::Base`)

rails-dev-tweaks inserting its autoload path probably caused rails to order them differently overall, but I don't think this is a problem in rails-dev-tweaks (could be wrong)

**Comment by gmoore** [comment]
I got really excited because I thought that Fa::Base would work (instead of FA::Base).  No luck, though.

I'll make those changes to the autoload paths and get back to this ticket tomorrow.  Thanks much for the help.

**Comment by nevir** [comment]
Yeah, it's a bit goofy around naming - `Fa::Base` should work too if it's operating properly!

**Comment by gmoore** [comment]
Our unorthodox autoload paths were so we could organize models in subfolders and not namespace them.  Some of those models are now involved in STI so we'd like to keep it that way if possible.

Anyway, I think the root of this problem came down to a mis-configured Gemfile wherein lots of spec dependencies (rspec, shoulda, etc...) were being loaded in development.  When I sorted that out, I got a Sprokets error which pointed to a mis-configured asset pipeline which indeed was the case.  We're in the middle of the upgrade now and the asset pipeline is a mess.

Once that gets sorted I'll re-introduce rails-dev-tweaks and go from there.  Will report back if any new problems.  Thanks very much!
